File: 2015/Day3/src/day3.py
import sys

direction = {'<':(-1, 0), '^':(0, 1), '>':(1, 0), 'v':(0, -1)}
debug = False

class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __repr__(self):
        return f"({self.x}, {self.y})"

def pos_found(visited_dict, pos):
    for value, key in enumerate(visited_dict):
        if(pos.x == key.x and pos.y == key.y):
            return key
    return None

def find_num_visited(commands : list) -> int:
    old_pos = Position(0, 0)
    visited_dict = {old_pos:1}
    for command in commands:
        pos = Position(0,0)
        pos.x = old_pos.x + direction[command][0]
        pos.y = old_pos.y + direction[command][1]
        key = pos_found(visited_dict, pos)
        if(key == None):
            visited_dict[pos] = 1
        else:
            visited_dict[key] += 1
        old_pos = pos
        if(debug):
            print(visited_dict)
    return len(visited_dict)

def find_num_visited_part2(commands : list) -> int:
    old_santa_pos = Position(0, 0)
    old_robot_pos = Position(0, 0)
    visited_dict = {old_santa_pos:1}
    for i in range(0, len(commands), 2):
        santa_pos = Position(0,0)
        santa_pos.x = old_santa_pos.x + direction[commands[i]][0]
        santa_pos.y = old_santa_pos.y + direction[commands[i]][1]
        key = pos_found(visited_dict, santa_pos)
        if(key == None):
            visited_dict[santa_pos] = 1
        else:
            visited_dict[key] += 1
        old_santa_pos = santa_pos

        robot_pos = Position(0,0)
        robot_pos.x = old_robot_pos.x + direction[commands[i+1]][0]
        robot_pos.y = old_robot_pos.y + direction[commands[i+1]][1]
        key = pos_found(visited_dict, robot_pos)
        if(key == None):
            visited_dict[robot_pos] = 1
        else:
            visited_dict[key] += 1
        old_robot_pos = robot_pos
    return len(visited_dict)

if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("Usage: python3 day2.py input.txt")
        exit
    commands = []
    with open(sys.argv[1], "r") as fileobj:
        commands = list(fileobj.read())
    while('\n' in commands):
        commands.remove('\n')
    num_visited = find_num_visited(commands)
    print(f"answer to part 1 is {num_visited}")

    num_visited = find_num_visited_part2(commands)
    print(f"answer to part 2 is {num_visited}")

File: 2015/Day3/src/day3.cpp
#include <iostream>
#include <set>
#include <fstream>
#include <utility> 


using namespace std;

int day3_solver(ifstream * file, bool is_robot)
{
    char command;
    pair <int, int> santa_pos = make_pair(0, 0);
    pair <int, int> robot_pos = make_pair(0, 0);
    set < pair<int, int> > visited = {santa_pos};
    int counter = 0;
    while(*file >> command)
    {
        pair<int, int> pos = make_pair(get<0>(santa_pos), get<1>(santa_pos));
        if(is_robot && counter % 2 != 0)
        {
            pos = make_pair(get<0>(robot_pos), robot_pos.second);
        }
        switch(command)
        {
            case 'v':
                pos = make_pair(get<0>(pos), get<1>(pos) + 1); 
                break;
            case '<':
                pos = make_pair(get<0>(pos) - 1, get<1>(pos));
                break;
            case '^':
                pos = make_pair(get<0>(pos), get<1>(pos) - 1);
                break;
            case '>':
                pos = make_pair(get<0>(pos) + 1, get<1>(pos));
                break;
            default:
                cout << "unknown command: " << command << endl;
                break;
        }
        visited.insert(pos);
        if(is_robot && counter % 2 != 0)
        {
            robot_pos = make_pair(get<0>(pos), get<1>(pos));
        }
        else
        {
            santa_pos = make_pair(get<0>(pos), get<1>(pos));
        }
        counter++;
   }
   return visited.size(); 
}


int main(int argc, char * argv[])
{
    if(argc != 2)
    {
        cout << "Usage: ./day3 input.txt\n";
        exit(1);
    }
    ifstream file(argv[1]);
    int part1_ans = day3_solver(&file, false);
    cout << "The answer to part 1 is " << part1_ans << endl;
    file.close();
    file.open(argv[1]);
    int part2_ans = day3_solver(&file, true);
    cout << "The answer to part 2 is " << part2_ans << endl;
    file.close();
    return 0;
}
